metadata/metadata.go

```diff
@@ -0,0 +1,145 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package metadata
+
+import (
+	"github.com/m3db/m3metrics/aggregation"
+	"github.com/m3db/m3metrics/op/applied"
+	"github.com/m3db/m3metrics/policy"
+)
+
+var (
+	// DefaultStagedMetadata represents a default staged metadata.
+	DefaultStagedMetadata StagedMetadata
+)
+
+// PipelineType describes the type of a pipeline.
+type PipelineType int
+
+// A list of supported pipeline types.
+const (
+	// A standard pipeline is a full pipeline that has not been processed.
+	// The first step of a standard pipeline is always an aggregation step.
+	// Metrics associated with a standard pipeline are raw, unaggregated metrics.
+	StandardType PipelineType = iota
+
+	// A forwarding pipeline is a sub-pipeline whose previous steps have
+	// been processed. There are no aggregation steps in a forwarding pipeline.
+	// Metrics associated with a forwarding pipeline are produced from
+	// previous steps of the same forwarding pipeline.
+	ForwardingType
+)
+
+// AggregationMetadata dictates how metrics should be aggregated.
+type AggregationMetadata struct {
+	// List of aggregation types.
+	AggregationID aggregation.ID
+
+	// List of storage policies.
+	StoragePolicies []policy.StoragePolicy
+}
+
+// IsDefault returns whether this is the default aggregation metadata.
+func (m AggregationMetadata) IsDefault() bool {
+	return m.AggregationID.IsDefault() && policy.IsDefaultStoragePolicies(m.StoragePolicies)
+}
+
+// StandardPipelineMetadata contains standard pipeline metadata.
+type StandardPipelineMetadata struct {
+	AggregationMetadata
+	applied.Pipeline
+}
+
+// IsDefault returns whether this is the default standard pipeline metadata.
+func (sm StandardPipelineMetadata) IsDefault() bool {
+	return sm.AggregationMetadata.IsDefault() && sm.Pipeline.IsEmpty()
+}
+
+// ForwardingPipelineMetadata contains forwarding pipeline metadata.
+type ForwardingPipelineMetadata struct {
+	applied.Pipeline
+}
+
+// IsDefault returns whether this is the default forwarding pipeline metadata.
+func (fm ForwardingPipelineMetadata) IsDefault() bool {
+	return fm.Pipeline.IsEmpty()
+}
+
+// PipelineMetadataUnion is a union of different pipeline metadatas.
+type PipelineMetadataUnion struct {
+	Type       PipelineType
+	Standard   []StandardPipelineMetadata
+	Forwarding ForwardingPipelineMetadata
+}
+
+// IsDefault returns whether this is the default pipeline metadata.
+func (pu PipelineMetadataUnion) IsDefault() bool {
+	return (pu.Type == StandardType && len(pu.Standard) == 0) ||
+		(pu.Type == ForwardingType && pu.Forwarding.IsDefault())
+}
+
+// Metadata represents the metadata associated with a metric.
+type Metadata struct {
+	AggregationMetadata
```

Comment:
Do we need AggregationMetadata here? What's the relation between the AggregationMetadata here and the AggregationMetadata in the StandardPipelineMetadatas, does this mean we will always have a AggregationMetadata here even when the PipelineMetadataUnion is a forwarding one?

Reply:
Yeah I updated the data structures here to make it much simpler.